### ISSUE TITLE
AMP Checkpointing

### ISSUE THREAD

**Issue by sahewat** [NVIDIA/apex#103]
Assuming it differs from normal pytorch usage, would it be possible to provide an example of the steps required to save and load model checkpoints with amp? Are there any specific considerations to take into account (especially when using DDP)?

**Comment by matthew-z** [comment]
I think there is nothing specific  you need to do for checkpointing with AMP.
For DDP, you only need to modify the code to do checkpointing only on worker0 (i.e., when rank==0)

**Comment by mcarilli** [comment]
Amp operates dynamically as your model is run.  It should not affect the way your model is saved/restored.  The lines of code that save and restore your model with Amp should be the same as the lines of code that save ([construct dict](https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_fp16_optimizer.py#L241-L247) + [call torch.save](https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_fp16_optimizer.py#L418-L421)) and [restore](https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main_fp16_optimizer.py#L164-L179) your model in the FP16_Optimizer example.

@matthew-z Do you mean for eval, or during training?

**Comment by matthew-z** [comment]
@mcarilli Thank you for the reply.  I think I should not ask here as it is more like a pytorch question so I deleted it. 

I mean both of them (training and validation), but the values like loss and BLEU scores are only for logging purpose.

**Comment by mcarilli** [comment]
Typically we manually reduce and average the metrics across processes, [for example](https://github.com/NVIDIA/apex/blob/master/examples/imagenet/main.py#L325) 

In our example, we allreduce the loss every iteration for averaging purposes, but in general, it's better to only do it every 10-ish iterations if you can help it, because the allreduce is expensive.  Also, you should only print the loss every 10-ish iterations if you can help it, because printing the loss incurs a host-device sync*.

*this is not a distributed thing, this is something that occurs within each process.

**Comment by matthew-z** [comment]
@mcarilli  Thank you for your detailed explanation!  I will try it!

**Comment by shoutOutYangJie** [comment]
> I think there is nothing specific you need to do for checkpointing with AMP.
> For DDP, you only need to modify the code to do checkpointing only on worker0 (i.e., when rank==0)

so how to ensure GPU0 saves checkpoint rather than other GPUs. I meet a question, 
You can find it [here](https://github.com/NVIDIA/apex/issues/629)
